### DISCUSSION TITLE
Merge vs Rebase

### DISCUSSION THREAD

**Discussion by nevack** [transmission/transmission#3188] [General]
One question I have after some experience of contributing to the project.

Why do you prefer to merge `main` into feature branch?
IMHO history looks cleaner if you rebase feature branch onto `main`.

CC @ckerr

**Comment by ile6695** [comment]
When merging a pull request, there is only one commit in the `main`.
When merging `main` into feature branch, rebase does look cleaner, but if I recall correctly it requires force pushing to the feature branch.

**Comment by ckerr** [comment]
I almost always squash-merge when merging PRs to main, so it's really whatever you prefer.

Whether you update your PR by merging or by rebasing, it will all look the same in `main` when all is said and done.

**Reply by nevack** [reply]
Yes, I do know that.
What I'm really asking is why do you prefer to make changes in user feature branches by merging `main` into them from time to time?
Like there https://github.com/transmission/transmission/pull/2870/commits and there https://github.com/transmission/transmission/pull/3183/commits

My main point is such merges require to keep local changes in sync with remote.
For example, if I commit some changes into local copy, I will be required to fix divergence before pushing.

This is why I prefer to rebase and force-push manually.
Ability to merge is already checked by Github (Conflicts), ability to build should be checked on CI by rebasing branch onto HEAD of target branch. Merge-pin or merge-ref, call it like you want.

**Comment by ckerr** [comment]
Hmm, you know what, I think what happened is  I had the "Update branch" button in GitHub set to "Update with merge commit" instead of "Update with rebase." :roll_eyes:  It doesn't have any policy significance.
